---
title: "Predicting e-features"
author: "Jiawei Huang"
output: html_document
---


#read data

```{r}
ef = read.csv("../data/e_features_filtered.csv",header = T,stringsAsFactors = F)
gf = read.csv("../data/geneExp_filtered.csv",header = T)
diff_ef = read.csv("../data/diff_ef.csv",header = T,stringsAsFactors = F)
diff_genes = read.csv("../data/diff_genes.csv",header = T,stringsAsFactors = F)
clustering = read.csv("../data/efeatures_NMA.csv",header = T,stringsAsFactors = F)
match = read.csv('../data/meta_NMA.csv',header=T,check.names = F,stringsAsFactors = F)
rownames(gf) = gf[,1];gf = gf[,-1]
rownames(ef) = colnames(gf);
cellnames = rownames(ef)
#t_type = sapply(cellnames,FUN = function(x){match$genotype[match$transcriptomics_sample_id==x]})
n = nrow(ef)
```

# predict

cluster 1 upstroke-downstroke-ratio-long-square 

```{r}
library(ggplot2)
Pred_R2 = c()
train_R2 = c()
set.seed(1)
train_ind = sample(1:n,n*0.9,replace = FALSE)
for (j in 2:5){
  pred_genes = data.frame(log10(t(gf[diff_genes$gene[diff_genes$cluster == paste0("Cluster ",j) & diff_genes$p_val_adj<0.05],])+1))
  pred_ef = ef[,diff_ef$diff_e_feature[diff_ef$cluster == j]]
  for (i in 1:ncol(pred_ef)){
    feature = colnames(pred_ef)[i]
    pred = cbind(pred_genes,pred_ef[,i])
    colnames(pred)[length(colnames(pred))] = "y"
    train_pred = pred[train_ind,]
    test_pred = pred[-train_ind,]
    fiti = lm(y~.,data = pred)
    Pred_R2i = 1- sum((predict.lm(fiti,test_pred)-test_pred$y)^2)/sum((test_pred$y - mean(test_pred$y))^2)
    Pred_R2 = c(Pred_R2,Pred_R2i)
    train_R2 = c(train_R2,(summary(fiti)$r.squared))
    #plot
    pred_df <- data.frame(x = predict.lm(fiti,test_pred),y = test_pred$y)
    lim_range = -min(pred_df)+max(pred_df)
    scatter = ggplot(pred_df, aes(x, y))+geom_point(col = "#0C2D48",alpha = 0.6,size = 2)+xlim(min(pred_df),max(pred_df)) + 
      ylim(min(pred_df),max(pred_df)) + theme_minimal() + 
      labs(y="observed",x="predicted")
    ggsave(file=paste0("../figure/pred/",feature,".png"), scatter,width = 10,height = 10, units = "cm")
  }
}
pred_res = data.frame(
  e_feature = diff_ef$diff_e_feature,
  cluster = diff_ef$cluster,
  p_val_adj = diff_ef$p_val_adj,
  Pred_R2 = Pred_R2,
  train_R2 = train_R2
)
write.csv(pred_res,"../data/pred_res.csv",row.names=F)
```

```{r}
pred_res$e_feature = sapply(pred_res$e_feature, gsub, pattern = ".", replacement = " ", fixed = TRUE)
library(RColorBrewer)
p<-ggplot(data=pred_res[!duplicated(pred_res$e_feature) & pred_res$Pred_R2>0.5,], 
          aes(x = reorder(e_feature,Pred_R2), y=Pred_R2)) + labs(y="enrichment")+
  geom_bar(stat="identity",width=0.4,fill = brewer.pal(6,"Spectral")[5],alpha = 0.8)+ theme_minimal() + 
  labs(y="test R square (>0.5)",x="") +
  coord_flip() + 
  theme(axis.text = element_text(size=12.5,lineheight=.8),
        axis.title = element_text(size=11,face = "bold")) 
p
ggsave(file=paste0("../figure/pred_R.png"), p)
```


