---
title: "Predicting e-features"
author: "Jiawei Huang"
output: html_document
---


#read data

```{r}
ef = read.csv("../data/efeature_filtered.csv",header = T,stringsAsFactors = F)
gf = read.csv("../data/expMat_filtered.csv",header = T)
diff_ef = read.csv("../data/diff_ef.csv",header = T,stringsAsFactors = F)
diff_genes = read.csv("../data/diff_genes.csv",header = T,stringsAsFactors = F)
clustering = read.csv("../data/efeatures_NMA.csv",header = T,stringsAsFactors = F)
match = read.csv('../data/20200711_patchseq_metadata_mouse.csv',header=T,check.names = F,stringsAsFactors = F)
rownames(ef) = ef[,1];ef = ef[,-1]
rownames(gf) = gf[,1];gf = gf[,-1]
cellnames = rownames(ef)
match$genotype = sapply(match$t_type,function(x){na.omit(unlist(strsplit(x, " ")))[1]})
t_type = sapply(cellnames,FUN = function(x){match$genotype[match$transcriptomics_sample_id==x]})
n = nrow(ef)
```

# predict

cluster 1 upstroke-downstroke-ratio-long-square 

```{r}
library(ggplot2)
Pred_R2 = c()
train_R2 = c()
set.seed(1)
train_ind = sample(1:n,n*0.9,replace = FALSE)
for (j in 1:5){
  pred_genes = data.frame(log10(t(gf[diff_genes$gene[diff_genes$cluster == paste0("Cluster ",j) & diff_genes$p_val_adj <0.05],])+1))
  pred_ef = ef[,diff_ef$diff_e_feature[diff_ef$cluster == j]]
  for (i in 1:ncol(pred_ef)){
    feature = colnames(pred_ef)[i]
    pred = cbind(pred_genes,pred_ef[,i])
    colnames(pred)[length(colnames(pred))] = "y"
    train_pred = pred[train_ind,]
    test_pred = pred[-train_ind,]
    fiti = lm(y~.,data = pred)
    Pred_R2i = 1- sum((predict.lm(fiti,test_pred)-test_pred$y)^2)/sum((test_pred$y - mean(test_pred$y))^2)
    Pred_R2 = c(Pred_R2,Pred_R2i)
    train_R2 = c(train_R2,(summary(fiti)$r.squared))
    #plot
    pred_df <- data.frame(x = predict.lm(fiti,test_pred),y = test_pred$y)
    lim_range = -min(pred_df)+max(pred_df)
    scatter = ggplot(pred_df, aes(x, y))+geom_point(col = "#0C2D48",alpha = 0.6,size = 2)+xlim(min(pred_df),max(pred_df)) + 
      ylim(min(pred_df),max(pred_df)) + theme_minimal() + 
      labs(y="observed",x="predicted")
    ggsave(file=paste0("../figure/pred/",feature,".png"), scatter,width = 10,height = 10, units = "cm")
  }
}
pred_res = data.frame(
  e_feature = diff_ef$diff_e_feature,
  cluster = diff_ef$cluster,
  p_val_adj = diff_ef$p_val_adj,
  Pred_R2 = Pred_R2,
  train_R2 = train_R2
)
write.csv(pred_res,"../data/pred_res.csv",row.names=F)
```

```{r}
pred_res$e_feature = sapply(pred_res$e_feature, gsub, pattern = "_", replacement = " ", fixed = TRUE)
library(RColorBrewer)
p<-ggplot(data=pred_res[!duplicated(pred_res$e_feature) & pred_res$Pred_R2>0.5,], 
          aes(x = reorder(e_feature,Pred_R2), y=Pred_R2)) + labs(y="enrichment")+
  geom_bar(stat="identity",width=0.4,fill = brewer.pal(6,"Spectral")[5],alpha = 0.8)+ theme_minimal() + 
  labs(y="test R square (>0.5)",x="") +
  coord_flip() + 
  theme(axis.text = element_text(size=12.5,lineheight=.8),
        axis.title = element_text(size=11,face = "bold")) 
p
ggsave(file=paste0("../figure/pred_R.png"), p)
```



```{r}
library(viridis)
library(plyr)
pal = brewer.pal(6,"Dark2")[1:5]
n = nrow(pred_ef)
pred_genes = data.frame(log10(t(gf[diff_genes$gene[diff_genes$cluster == paste0("Cluster ",1)],])+1))
pred = cbind(pred_genes,pred_ef[,1])
colnames(pred)[length(colnames(pred))] = "y"
train_pred = pred[train_ind,]
test_pred = pred[-train_ind,]
fit = lm(y~.,data = train_pred)
plot(fit$fitted.values,train_pred$y)
plot(predict.lm(fit,test_pred),test_pred$y,
     col = mapvalues(clustering$gmm_cluster,1:5,pal),
     ylab = "predicted U/D ratio (long square)",xlab = "Measured U/D ratio (long square)",pch = 19)
legend("topleft", pch = 19, title="GMM Clusters",
   paste0("Cluster ", 1:5), col=pal, cex=0.8)
```

```{r}
library(ggplot2)
library(gridExtra)
pred_df <- data.frame(x = predict.lm(fit,test_pred),y = test_pred$y)
lim_range = -min(pred_df)+max(pred_df)
hist_top <- ggplot() + geom_histogram(aes(pred_df$x),bins = 10, col = "gray") + theme_minimal()+labs(y="",x="")
empty <- ggplot()+geom_point(aes(1,1), colour="white")+
         theme(axis.ticks=element_blank(), 
               panel.background=element_blank(), 
               axis.text.x=element_blank(), axis.text.y=element_blank(),           
               axis.title.x=element_blank(), axis.title.y=element_blank())

scatter <- ggplot(pred_df, aes(x, y))+geom_point()+xlim(min(pred_df),max(pred_df)) + 
  ylim(min(pred_df),max(pred_df)) + theme_minimal() + 
  labs(y="observed",x="predicted")
hist_right <- ggplot()+geom_histogram(aes(pred_df$y),bins = 10, col = "gray", fill = "#6D1A36")+coord_flip()+ theme_minimal()+labs(y="",x="")
grid.arrange(hist_top, empty, scatter, hist_right, ncol=2, nrow=2, widths=c(4, 1), heights=c(1, 4))
```

